bentoml/metrics.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import typing as t
+import logging
+from typing import TYPE_CHECKING
+
+from simple_di import inject
+from simple_di import Provide
+
+logger = logging.getLogger(__name__)
+
+from ._internal.configuration.containers import BentoMLContainer
+
+if TYPE_CHECKING:
+    from bentoml._internal.server.metrics.prometheus import PrometheusClient
+
+
+class BaseMetric(object):
+    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
+        if "registry" in kwargs:
+            kwargs.pop("registry")
+            logger.warning("registry cannot be in passed as argument")
+        self._args = args
+        self._kwargs = kwargs
+        self._metric = None
+
+    @inject
+    def _set_up_metric(
+        self, metric_client: PrometheusClient = Provide[BentoMLContainer.metrics_client]
+    ) -> None:
+        cls = type(self).__name__
+        raise Exception(f"{cls} has not implemented a set_up method")
```

Comment:
Should be an abstractmethod then.
```suggestion
    @abstractmethod
    def _setup_metric(self, metric_client: PrometheusClient) -> None:
        raise NotImplementedError
```